### ISSUE TITLE
Investigate React starter kit

### ISSUE THREAD

**Issue by ericelliott** [ericelliott/universal-react-boilerplate#17]
Is there anything useful in Rick Wong's [React Isomorphic Starterkit](https://github.com/RickWong/react-isomorphic-starterkit)?

Looks to me like it doesn't do much but render a single React view with Hapi but I didn't try it out or investigate very deeply.

Looks like our project is much more detailed and covers a whole lot more ground. Take a peek.. let me know if there's anything worth emulating.

**Comment by zebulonj** [comment]
@ericelliott I'll take a look at this. I've taken the plunge into React in a test project of my own, and am happy with what is coming of it. A key piece of effectively utilizing React is delivering the server rendered markup in a way that React can effectively bind to it on the client-side. I think I've established a solid pattern for doing this. Incidentally, that delivery for re-binding dovetails with the isomorphic routing. I see in your roadmap that you want the router to issue commands rather than render views. If you could elaborate on that a little, I might be able to incorporate that approach into what I'm doing with React and share it here.

**Comment by ericelliott** [comment]
Yeah, the idea is to invoke a command on a command manager (not directly from the view, of course). The command manager is essentially a history of commands in the form of an immutable trie data structure so that it has full support for history / undo / redo / replay, etc...

Commands can be issued in the browser and automatically stream back to the server.

The effect of a command is to actually make a change to the app state's data models. Hopefully the app state will also be represented in immutable data structures.

References:

* [Immutable Data and React](https://www.youtube.com/watch?v=I7IdS-PbEgI&list=PLb0IAmt7-GS1cbw4qonlQztYV1TAW0sCr)
* [Immutability: Putting the Dream Machine to Work](https://www.youtube.com/watch?v=SiFwRtCnxv4)
* [Asynchronous Programming at Netflix](https://www.youtube.com/watch?v=gawmdhCNy-A)

And for a general overview of why all this is important: [The Two Pillars of JavaScript Pt 2](https://medium.com/javascript-scene/the-two-pillars-of-javascript-pt-2-functional-programming-a63aa53a41a4)

Command managers have a rich history, too. I came to them through computer game programming. Game programming is interesting because writing to the graphics card has always been too slow for realtime screen drawing, so we've been using virtual screens to buffer changes and then swapping the buffer to screen all at once. Basically a pixel-level version of the React virtual DOM (usually without the cool diffing, though... different era).

But another interesting thing that a lot of game engines do is they manipulate the virtual world with commands, and (sometimes) keep a history of those commands for features like auto-save game state. Some of them even feature rewind / replay functionality.

In some game engines, you can pull up a command console and issue commands with a REPL-like CLI.

Basically, none of these ideas are new, they're just not common in web development, yet.

**Comment by ericelliott** [comment]
I believe I mentioned the command pattern as a glancing reference in Appendix A of my book:

> Would it be useful to log the cases that get executed, for example, to create an undo/redo stack, or log user actions to your servers for analysis? (command manager)

but then I never really explained what a command manager is. I still haven't made a detailed writeup. Let me know if you find one we can link to for reference.

**Comment by zebulonj** [comment]
This one isn't bad, in that it provides fairly digestible examples: http://gameprogrammingpatterns.com/command.html. That said, I think the subject merits discussion in the context of this specific use case.

**Comment by RickWong** [comment]
I've updated my react-isomorphic-starterkit to include new stuff like React-Async for true asynchronous serverside rendering with React. Also server & client hot updates are both near instant. That's awesomely productive.

Another thing I've created is React Transmit. It's a Relay-inspired library that pre-populates props before rendering using Relay syntax but instead of GraphQL it allows multipurpose Promises. Check it out, you might fall in love with it: https://github.com/RickWong/react-transmit I'm just might replace React-Async with Transmit if I can make it work isomorphically (without this.setState).

**Comment by ericelliott** [comment]
:+1:

**Comment by SOSANA** [comment]
Wow thanks for this open issue discussion on react.js. Great enlightenment reading. Curious what decisions if any has been made.

**Comment by ericelliott** [comment]
@SOSANA We haven't committed any ideas RE: React to the repo yet, but I'm experimenting.

On the subject of the command pattern, the Flux architecture is similar:

[The essence of Flux architecture explained](https://speakerdeck.com/jmorrell/jsconf-uy-flux-those-who-forget-the-past-dot-dot-dot)

If it's implemented with immutables implemented on tries, that's ideal:

[Make React faster with Immutable.js stores](https://facebook.github.io/react/docs/advanced-performance.html)

**Comment by SOSANA** [comment]
Stumbled upon this other week reactjs/express-react-views, an Express view engine which renders React components on server. It renders static markup and *does not* support mounting those views on the client https://github.com/reactjs/express-react-views

**Comment by ericelliott** [comment]
Have you seen PayPal's react-engine?

**Comment by zebulonj** [comment]
Just saw a tweet about it this morning. Was going to take a look. What do you think of it?




—

**Comment by ericelliott** [comment]
I'm experimenting with `react-engine` right now. I'll update when I've had a chance to live with it for a while.

**Comment by ericelliott** [comment]
Here's a React-friendly diagram of the command pattern:

![command-pattern-diagram](https://cloud.githubusercontent.com/assets/364727/7505705/054e80f2-f413-11e4-95fa-0b432f1ba262.png)

**Comment by SOSANA** [comment]
@ericelliott  man the other night it hit me. Sounds crazy but had this idea, hit my like that scene from back to the future where doc hits his head on the bathroom sink and conceived the idea for the flux capacitor lol Totally have to do the MERN stack (mongodb, expressjs, react.js, node.js), the isomorphic stack.

**Comment by ericelliott** [comment]
@SOSANA That's cool! I want this particular boilerplate to be db-agnostic (which is why I want to use object-list for the store), but I do want a MongoDB adaptor for object-list. =)

Perf-wise, I think for the **single store pattern** (where all record types are stored in one root collection and you look up different record types by querying for keys) Redis may be more performant, but if the db was causing a bottleneck I would definitely benchmark that instead of just assuming.

**Comment by SOSANA** [comment]
@ericelliott wow man check this out...faster than react and it comes with its own optional mvc, models simialr to backbone https://github.com/astoilkov/jsblocks

**Comment by ericelliott** [comment]
Looks cool for tiny prototypes and hackathons. The 2-way data binding will cause headaches in large / complex apps.
